sw/device/lib/dif/dif_pwrmgr.c

```diff
@@ -4,8 +4,543 @@
+                                        reg_info.write_enable_reg_offset);
+  // Locked if write enable bit is set to 0.
+  return !bitfield_bit32_read(reg_val, reg_info.write_enable_bit_index);
+}
+
+dif_pwrmgr_result_t dif_pwrmgr_init(dif_pwrmgr_params_t params,
+                                    dif_pwrmgr_t *pwrmgr) {
+  if (pwrmgr == NULL) {
+    return kDifPwrmgrBadArg;
+  }
+
+  *pwrmgr = (dif_pwrmgr_t){.params = params};
+
+  return kDifPwrmgrOk;
+}
+
+dif_pwrmgr_config_result_t dif_pwrmgr_low_power_set_enabled(
+    const dif_pwrmgr_t *pwrmgr, dif_pwrmgr_toggle_t new_state) {
+  if (pwrmgr == NULL || !is_valid_toggle(new_state)) {
+    return kDifPwrmgrConfigResultBadArg;
+  }
+
+  if (control_register_is_locked(pwrmgr)) {
+    return kDifPwrMgrConfigResultLocked;
+  }
+
+  // Read-modify-write.
+  uint32_t reg_val =
+      mmio_region_read32(pwrmgr->params.base_addr, PWRMGR_CONTROL_REG_OFFSET);
+  reg_val = bitfield_bit32_write(reg_val, PWRMGR_CONTROL_LOW_POWER_HINT,
+                                 new_state == kDifPwrmgrToggleEnabled);
+  mmio_region_write32(pwrmgr->params.base_addr, PWRMGR_CONTROL_REG_OFFSET,
+                      reg_val);
+
+  // Sync changes to slow clock domain.
+  sync_slow_clock_domain(pwrmgr);
+
+  return kDifPwrmgrConfigResultOk;
+}
+
+dif_pwrmgr_result_t dif_pwrmgr_low_power_get_enabled(
+    const dif_pwrmgr_t *pwrmgr, dif_pwrmgr_toggle_t *cur_state) {
+  if (pwrmgr == NULL || cur_state == NULL) {
+    return kDifPwrmgrBadArg;
+  }
+
+  uint32_t reg_val =
+      mmio_region_read32(pwrmgr->params.base_addr, PWRMGR_CONTROL_REG_OFFSET);
+  *cur_state = bitfield_bit32_read(reg_val, PWRMGR_CONTROL_LOW_POWER_HINT)
+                   ? kDifPwrmgrToggleEnabled
+                   : kDifPwrmgrToggleDisabled;
+
+  return kDifPwrmgrOk;
+}
+
+dif_pwrmgr_config_result_t dif_pwrmgr_set_domain_config(
+    const dif_pwrmgr_t *pwrmgr, dif_pwrmgr_domain_config_t config) {
+  if (pwrmgr == NULL || !is_valid_for_bitfield(config, kDomainConfigBitfield)) {
+    return kDifPwrmgrConfigResultBadArg;
+  }
+
+  if (control_register_is_locked(pwrmgr)) {
+    return kDifPwrMgrConfigResultLocked;
+  }
+
+  // Read-modify-write.
+  uint32_t reg_val =
+      mmio_region_read32(pwrmgr->params.base_addr, PWRMGR_CONTROL_REG_OFFSET);
+  reg_val = bitfield_field32_write(reg_val, kDomainConfigBitfield, config);
+  mmio_region_write32(pwrmgr->params.base_addr, PWRMGR_CONTROL_REG_OFFSET,
+                      reg_val);
+
+  // Sync changes to slow clock domain.
+  sync_slow_clock_domain(pwrmgr);
+
+  return kDifPwrmgrConfigResultOk;
+}
+
+dif_pwrmgr_result_t dif_pwrmgr_get_domain_config(
+    const dif_pwrmgr_t *pwrmgr, dif_pwrmgr_domain_config_t *config) {
+  if (pwrmgr == NULL || config == NULL) {
+    return kDifPwrmgrBadArg;
+  }
+
+  uint32_t reg_val =
+      mmio_region_read32(pwrmgr->params.base_addr, PWRMGR_CONTROL_REG_OFFSET);
+  *config = bitfield_field32_read(reg_val, kDomainConfigBitfield);
```

Comment:
Call me weird, but what I honestly would have done here is write something like
```c
uint32_t reg = read32();
*config = 0;

// Repeat for each config bit.
*config = bitfield_bit32_write(*config, kConfigBit0, bitfield_bit32_read(reg, CONFIG_BIT_0));
```

I think this style is slightly clearer, and the compiler will see through it and make the bit shuffling efficient (or a no-op, if possible).

(Note that, in this case, kConfigBit0 is no longer a power of two but a bit index.)

Reply:
This is related to the values of the enum constants. In this case, IMHO using bitmasks is leaner. We can always revisit this if we decide to enforce bit indeces.